bike.js

```diff
@@ -0,0 +1,145 @@
+  message.raw.tags.push('#bike-distance-saved')
+}
+
+async function _hasUserBikedToWorkBefore(state, userId) {
+  const rows = await state.db.query('SELECT COUNT(*) FROM bike_runs WHERE user_id = $1', [userId + ''])
+  return parseInt(rows[0].count, 10) > 0
+}
+
+async function _handleDistanceInputAbsentForBikeDone(message, state, hasBikedToWorkBefore) {
+  if (hasBikedToWorkBefore) {
+    // Regular ride
+    return await _getUserBikedToWorkDistance(state, message.author.id)
+  }
+  // First ride
+  await message.send('You haven\'t biked to work yet, please put the number of km you ride to work after the hashtag.')
+  return null
+}
+
+async function _handleDistanceInputPresentForBikeDone(message, state, hasBikedToWorkBefore, distanceMeters) {
+  if (!hasBikedToWorkBefore) {
+    // First ride
+    await _setUserBikedToWorkDistance(state, message, distanceMeters)
+    await message.send('This is your first ride! Your input distance has been saved as your regular biking distance.')
+  } else {
+    // Irregular ride, or the user wishes to save his regular bike distance
+    if (message.words.includes('#save')) {
+      await _setUserBikedToWorkDistance(state, message, distanceMeters)
+    }
+  }
+}
+
+async function _getBikeDistanceMetersToSave(message, state, distanceKm) {
+  const hasBikedToWorkBefore = await _hasUserBikedToWorkBefore(state, message.author.id)
+
+  let distanceMeters;
+  if (distanceKm === null) {
+    // No input of distance. Either user's first ride or regular ride
+    distanceMeters = await _handleDistanceInputAbsentForBikeDone(message, state, hasBikedToWorkBefore)
+  } else {
+    // Manual input of distance. Either user's first ride or irregular ride
+    distanceMeters = distanceKm * 1000
+    if (distanceMeters) await _handleDistanceInputPresentForBikeDone(message, state, hasBikedToWorkBefore, distanceMeters)
+  }
+  return distanceMeters
+}
+
+async function _saveBikeRun (message, state, distanceMeters) {
+  if (!distanceMeters) {
+    await message.tag('#no-distance-given')
+  } else {
+    // Save the run
+    await state.db.query('INSERT INTO bike_runs (user_id, distance) VALUES ($1, $2)', [message.author.id, distanceMeters])
+    await message.tag('#gg')
+    const rows = await state.db.query('SELECT SUM(distance) AS total FROM bike_runs')
+    const total = rows[0].total
+    const achievement = state.achievements.find(({ height }) => height > (total - distanceMeters) && height <= total)
+    if (achievement) {
+      await message.send(`@team, you just reached the ${achievement.name} (${achievement.location}), with a total of ${achievement.height / 1000} km!`)
+    }
+  }
+}
+
+async function onBikeDone (message, state) {
+  await state.db.query('INSERT INTO users (id, name) VALUES ($1, $2) ON CONFLICT (id) DO UPDATE SET name = $2', [message.author.id, message.author.name])
+
+  const number = message.words[message.words.indexOf(message.doneHash) + 1]
```

Comment:
What does `indexOf(message.doneHash) + 1` number represent?

Reply:
`indexOf(message.doneHash)` : Position of the hashtag word (e.g. `#btw`) in the array of words of the user message
`indexOf(message.doneHash) + 1`: Position of the optional word following the hashtag word, which we expect to be the number of kilometers biked when present.

See [here](https://github.com/busbud/stairs/pull/6/files#diff-168726dbe96b3ce427e7fedce31bb0bcR52)